crates/bevy_ecs/src/world/mod.rs

```diff
@@ -1691,13 +1702,23 @@ impl World {
         value: R,
         caller: MaybeLocation,
     ) {
-        let component_id = self.components_registrator().register_resource::<R>();
-        OwningPtr::make(value, |ptr| {
-            // SAFETY: component_id was just initialized and corresponds to resource of type R.
-            unsafe {
-                self.insert_resource_by_id(component_id, ptr, caller);
-            }
-        });
+        move_as_ptr!(value);
+        // if the resource already exists, we replace it on the same entity
+        if let Some(component_id) = self.resource_id::<R>()
```

Comment:
I think this needs to use `self.components_registrator().register_resource::<R>();` like the old code, or else it's possible that `world.insert_resource(R)` will fail to insert the resource because it wasn't previously initialized.